src/Conversion/ONNXToKrnl/Math/Gemm.cpp

```diff
@@ -20,181 +21,97 @@ struct ONNXGemmOpLowering : public ConversionPattern {
+    // Get shape.
     ONNXGemmOpAdaptor operandAdaptor(operands);
-    A = operandAdaptor.A();
-    B = operandAdaptor.B();
-    if (hasBias)
-      C = operandAdaptor.C();
-
-    auto memRefType = convertToMemRefType(*op->result_type_begin());
-
-    auto alphaAttr = FloatAttr::get(memRefType.getElementType(),
-        llvm::dyn_cast<GemmOp>(op).alpha().convertToFloat());
-    auto betaAttr = FloatAttr::get(memRefType.getElementType(),
-        llvm::dyn_cast<GemmOp>(op).beta().convertToFloat());
-    auto alpha = rewriter.create<ConstantOp>(loc, alphaAttr);
-    auto beta = rewriter.create<ConstantOp>(loc, betaAttr);
-
-    bool isTransA = (llvm::dyn_cast<GemmOp>(op).transA() != 0);
-    bool isTransB = (llvm::dyn_cast<GemmOp>(op).transB() != 0);
-
-    // Insert an allocation and deallocation for the result of this operation.
-    Value alloc;
-    bool insertDealloc = checkInsertDealloc(op);
-    if (hasAllConstantDimensions(memRefType))
-      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
-    else {
-      auto memRefShape = memRefType.getShape();
-      SmallVector<Value, 2> allocOperands;
-      if (memRefShape[0] < 0) {
-        auto dim = rewriter.create<DimOp>(loc, A, (isTransA) ? 1 : 0);
-        allocOperands.emplace_back(dim);
+    ONNXGemmOp gemmOp = llvm::cast<ONNXGemmOp>(op);
+    Location loc = op->getLoc();
+    ONNXGemmOpShapeHelper shapeHelper(&gemmOp, &rewriter);
+    assert(succeeded(shapeHelper.Compute(operandAdaptor)));
+    IndexExprContext outerContext(shapeHelper.context);
+
+    // Insert an allocation and deallocation for the output of this operation.
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    Type elementType = outputMemRefType.getElementType();
+    Value alloc = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.outputDims);
+
+    // Get the constants: zero, alpha,and beta.
+    float alphaLit = gemmOp.alpha().convertToFloat();
+    float betaLit = gemmOp.beta().convertToFloat();
+    Value alpha = emitConstantOp(rewriter, loc, elementType, alphaLit);
+    Value beta = emitConstantOp(rewriter, loc, elementType, betaLit);
+    Value zero = emitConstantOp(rewriter, loc, elementType, 0);
+
+    // loop iterations N=0 & M-1 going over each of the res[n, m] values.
+    BuildKrnlLoop outputLoops(rewriter, loc, 2);
+    outputLoops.createDefineOp();
+    outputLoops.pushAllBounds(shapeHelper.outputDims);
+    outputLoops.createIterateOp();
+    rewriter.setInsertionPointToStart(outputLoops.getIterateBlock());
+
+    // Compute the access functions for res[n,m].
+    IndexExpr n =
+        outerContext.createLoopIterIndex(outputLoops.getInductionVar(0));
+    IndexExpr m =
+        outerContext.createLoopIterIndex(outputLoops.getInductionVar(1));
+    SmallVector<IndexExpr, 4> resAccessFct({n, m});
+
+    // Insert res[n,m] = 0.
+    // rewriter.create<AffineStoreOp>(loc, zero, alloc, resAccessFct);
```

Comment:
Debug commment ?

Reply:
Actually it looks like it's needed based on the comment above.

Reply:
I believe I need that init back. Have added it back.

Reply:
It was needed, it was a bug not to have it. Tian confirmed it.

Reply:
Nice!